R/derive_date_vars.R

```diff
@@ -101,214 +117,730 @@
     }
-  } else {
-    # no imputation
-    imputed_date <- if_else(n_chr >= 10 & valid_dtc, substr(dtc, 1, 10), NA_character_)
   }
+  # Determine target components ----
+  target <- get_imputation_target_date(
+    date_imputation = date_imputation,
+    month = partial[["month"]]
+  )
 
-  if (!is.null(time_imputation)) {
-    # impute time
-    assert_that(is_valid_time_entry(time_imputation))
-    # make keywords case-insensitive
-    time_imputation <- str_to_lower(time_imputation)
-    if (time_imputation == "first") {
-      imputed_time <- "00:00:00"
-      sec <- ":00"
-      min <- ":00"
-      h <- "00"
-    } else if (time_imputation == "last") {
-      imputed_time <- "23:59:59"
-      sec <- ":59"
-      min <- ":59"
-      h <- "23"
-    } else {
-      imputed_time <- time_imputation
-      sec <- paste0(":", paste0(substr(dtc, 18, 19), substr(time_imputation, 7, 8)))
-      min <- paste0(":", paste0(substr(dtc, 15, 16), substr(time_imputation, 4, 5)))
-      h <- paste0(substr(dtc, 12, 13), substr(time_imputation, 1, 2))
+  for (c in components) {
+    if (highest_imputation < dt_level(imputation_levels[[c]])) {
+      target[[c]] <- "xx"
     }
+  }
+
+  # Impute ----
+  imputed <- vector("list", 3)
+  names(imputed) <- components
+  for (c in components) {
+    imputed[[c]] <- if_else(is.na(partial[[c]]), target[[c]], partial[[c]])
+  }
+
+  imputed_dtc <-
+    paste(imputed[["year"]], imputed[["month"]], imputed[["day"]], sep = "-")
 
-    imputed_time <- case_when(
-      n_chr >= 19 ~ substr(dtc, 12, 19),
-      n_chr == 16 ~ paste0(substr(dtc, 12, 16), sec),
-      n_chr == 13 ~ paste0(substr(dtc, 12, 13), min, sec),
-      n_chr == 10 ~ paste0(h, min, sec),
-      TRUE ~ imputed_time
+  imputed_dtc <-
+    if_else(
+      str_detect(imputed_dtc, "x"),
+      NA_character_,
+      imputed_dtc
     )
-  } else {
-    # no imputation
-    imputed_time <- if_else(n_chr >= 19 & valid_dtc, substr(dtc, 12, 19), NA_character_)
+
+  if (date_imputation == "last") {
+    imputed_dtc <-
+      if_else(
+        is.na(partial[["day"]]),
+        strftime(
+          rollback(ymd(imputed_dtc) + months(1)),
+          format = "%Y-%m-%d"
+        ),
+        imputed_dtc
+      )
   }
 
-  imputed_dtc <- if_else(
-    !is.na(imputed_date) & !is.na(imputed_time),
-    paste0(imputed_date, "T", imputed_time),
-    NA_character_
+  # Handle min_dates and max_dates argument ----
+  restrict_imputed_dtc_dt(
+    dtc,
+    imputed_dtc = imputed_dtc,
+    min_dates = min_dates,
+    max_dates = max_dates
   )
+}
+
+#' Create a `dt_level` object
+#'
+#' @param level Date level
+#'
+#'   *Permitted Values*: `"Y"` (year, highest level), `"M"` (month), `"D"`
+#'   (day), `"n"` (none, lowest level)
+#'
+#' @returns A `dt_level` object
+#'
+#' @details A `dt_level` object is an ordered factor, i.e., two objects can be
+#'   compared.
+#'
+#' @author Stefan Bundfuss
+dt_level <- function(level) {
```

Comment:
Maybe `get_dt_level` so all these helper functions have similar theme?

Reply:
The function returns an object of class `dt_level`. The convention is that the function has the same name as the class. Thus I would not change the name.